// @tamagui/toast@2.0.0-rc.36/dist/esm/Toast.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { AnimatePresence } from "@tamagui/animate-presence";
import { styled, useEvent } from "@tamagui/core";
import { composeEventHandlers, withStaticProperties } from "@tamagui/helpers";
import { YStack } from "@tamagui/stacks";
import { SizableText } from "@tamagui/text";
import { useControllableState } from "@tamagui/use-controllable-state";
import * as React from "react";
import { ToastAnnounceExclude } from "./ToastAnnounce.native.js";
import { useToast, useToastController, useToastState } from "./ToastImperative.native.js";
import { ToastImpl, ToastImplFrame, useToastInteractiveContext } from "./ToastImpl.native.js";
import { ToastProvider } from "./ToastProvider.native.js";
import { ToastViewport } from "./ToastViewport.native.js";
var TITLE_NAME = "ToastTitle",
  ToastTitle = styled(SizableText, {
    name: TITLE_NAME,
    variants: {
      unstyled: {
        false: {
          color: "$color",
          size: "$4"
        }
      }
    },
    defaultVariants: {
      unstyled: process.env.TAMAGUI_HEADLESS === "1"
    }
  }),
  DESCRIPTION_NAME = "ToastDescription",
  ToastDescription = styled(SizableText, {
    name: DESCRIPTION_NAME,
    variants: {
      unstyled: {
        false: {
          color: "$color11",
          size: "$1"
        }
      }
    },
    defaultVariants: {
      unstyled: process.env.TAMAGUI_HEADLESS === "1"
    }
  }),
  ACTION_NAME = "ToastAction",
  ToastAction = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
    var {
      altText,
      ...actionProps
    } = props;
    return altText ? /* @__PURE__ */_jsx(ToastAnnounceExclude, {
      altText,
      asChild: !0,
      children: /* @__PURE__ */_jsx(ToastClose, {
        ...actionProps,
        ref: forwardedRef
      })
    }) : null;
  });
ToastAction.propTypes = {
  altText(props) {
    return props.altText ? null : new Error(`Missing prop \`altText\` expected on \`${ACTION_NAME}\``);
  }
};
var CLOSE_NAME = "ToastClose",
  ToastCloseFrame = styled(YStack, {
    name: CLOSE_NAME,
    render: "button"
  }),
  ToastClose = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
    var {
        scope,
        ...closeProps
      } = props,
      interactiveContext = useToastInteractiveContext(scope);
    return /* @__PURE__ */_jsx(ToastAnnounceExclude, {
      asChild: !0,
      children: /* @__PURE__ */_jsx(ToastCloseFrame, {
        "aria-label": "Close",
        ...closeProps,
        ref: forwardedRef,
        onPress: composeEventHandlers(props.onPress, interactiveContext.onClose)
      })
    });
  }),
  ToastComponent = ToastImplFrame.styleable(function (props, forwardedRef) {
    var {
        forceMount,
        open: openProp,
        defaultOpen,
        onOpenChange,
        ...toastProps
      } = props,
      [open, setOpen] = useControllableState({
        prop: openProp,
        defaultProp: defaultOpen ?? !0,
        onChange: onOpenChange,
        strategy: "most-recent-wins"
      }),
      currentToast = useToastState(),
      {
        hide
      } = useToastController(),
      id = React.useId(),
      onPause = useEvent(props.onPause),
      onResume = useEvent(props.onResume),
      isHide = currentToast?.hide === !0,
      shouldShow = (forceMount || open) && !isHide;
    return /* @__PURE__ */_jsx(AnimatePresence, {
      children: shouldShow ? /* @__PURE__ */_jsx(ToastImpl, {
        id,
        open,
        ...toastProps,
        ref: forwardedRef,
        onClose: function () {
          setOpen(!1), hide();
        },
        onPause,
        onResume,
        onSwipeEnd: composeEventHandlers(props.onSwipeEnd, function (event) {
          setOpen(!1);
        })
      }) : null
    }, id);
  }),
  Toast2 = withStaticProperties(ToastComponent, {
    Title: ToastTitle,
    Description: ToastDescription,
    Action: ToastAction,
    Close: ToastClose
  });
export { Toast2 as Toast, ToastProvider, ToastViewport, useToast, useToastController, useToastState };
