// @tamagui/toast@2.0.0-rc.36/dist/esm/ToastAnnounce.native.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useIsomorphicLayoutEffect } from "@tamagui/constants";
import { View, Text, styled, useEvent } from "@tamagui/core";
import { Portal } from "@tamagui/portal";
import { startTransition } from "@tamagui/start-transition";
import { VisuallyHidden } from "@tamagui/visually-hidden";
import * as React from "react";
import { useToastProviderContext } from "./ToastProvider.native.js";
var ToastAnnounceExcludeFrame = styled(View, {
    name: "ToastAnnounceExclude"
  }),
  ToastAnnounceExclude = /* @__PURE__ */React.forwardRef(function (props, forwardedRef) {
    var {
      altText,
      ...announceExcludeProps
    } = props;
    return /* @__PURE__ */_jsx(ToastAnnounceExcludeFrame, {
      "data-toast-announce-exclude": "",
      "data-toast-announce-alt": altText || void 0,
      ...announceExcludeProps,
      ref: forwardedRef
    });
  }),
  ToastAnnounce = function (props) {
    var {
        scope,
        children,
        ...announceProps
      } = props,
      context = useToastProviderContext(scope),
      [renderAnnounceText, setRenderAnnounceText] = React.useState(!1),
      [isAnnounced, setIsAnnounced] = React.useState(!1);
    return useNextFrame(function () {
      startTransition(function () {
        setRenderAnnounceText(!0);
      });
    }), React.useEffect(function () {
      var timer = setTimeout(function () {
        return setIsAnnounced(!0);
      }, 1e3);
      return function () {
        return clearTimeout(timer);
      };
    }, []), isAnnounced ? null : /* @__PURE__ */_jsx(Portal, {
      children: /* @__PURE__ */_jsx(VisuallyHidden, {
        ...announceProps,
        children: renderAnnounceText && /* @__PURE__ */_jsxs(Text, {
          children: [context.label, " ", children]
        })
      })
    });
  };
function useNextFrame() {
  var callback = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : function () {},
    fn = useEvent(callback);
  useIsomorphicLayoutEffect(function () {
    var raf1 = 0,
      raf2 = 0;
    return raf1 = requestAnimationFrame(function () {
      raf2 = requestAnimationFrame(fn);
    }), function () {
      cancelAnimationFrame(raf1), cancelAnimationFrame(raf2);
    };
  }, [fn]);
}
export { ToastAnnounce, ToastAnnounceExclude };
