// @tamagui/react-native-web-internals@2.0.0-rc.36/dist/esm/modules/createDOMProps/index.mjs
import { getCSSStylesAtomic } from "@tamagui/web";
import { AccessibilityUtil } from "../AccessibilityUtil/index.mjs";
const emptyObject = {},
  hasOwnProperty = Object.prototype.hasOwnProperty,
  isArray = Array.isArray,
  reactNativeOnlyProps = {
    collapsable: !0,
    contentContainerStyle: !0,
    contentOffset: !0,
    decelerationRate: !0,
    maintainVisibleContentPosition: !0,
    onLayout: !0,
    onMomentumScrollBegin: !0,
    onMomentumScrollEnd: !0,
    onMoveShouldSetResponder: !0,
    onMoveShouldSetResponderCapture: !0,
    onResponderEnd: !0,
    onResponderGrant: !0,
    onResponderMove: !0,
    onResponderReject: !0,
    onResponderRelease: !0,
    onResponderStart: !0,
    onResponderTerminate: !0,
    onResponderTerminationRequest: !0,
    onScrollBeginDrag: !0,
    onScrollEndDrag: !0,
    onScrollShouldSetResponder: !0,
    onScrollShouldSetResponderCapture: !0,
    onSelectionChangeShouldSetResponder: !0,
    onSelectionChangeShouldSetResponderCapture: !0,
    onStartShouldSetResponder: !0,
    onStartShouldSetResponderCapture: !0,
    refreshControl: !0,
    removeClippedSubviews: !0,
    scrollEnabled: !0,
    scrollEventThrottle: !0,
    scrollIndicatorInsets: !0,
    showsHorizontalScrollIndicator: !0,
    showsVerticalScrollIndicator: !0,
    snapToAlignment: !0,
    snapToEnd: !0,
    snapToInterval: !0,
    snapToOffsets: !0,
    snapToStart: !0,
    stickyHeaderIndices: !0,
    ScrollComponent: !0
  },
  uppercasePattern = /[A-Z]/g;
function toHyphenLower(match) {
  return "-" + match.toLowerCase();
}
function hyphenateString(str) {
  return str.replace(uppercasePattern, toHyphenLower);
}
function processIDRefList(idRefList) {
  return isArray(idRefList) ? idRefList.join(" ") : idRefList;
}
function flattenStyle(style) {
  if (style === null || typeof style != "object") return;
  if (!isArray(style)) return style;
  const result = {};
  for (let i = 0, styleLength = style.length; i < styleLength; ++i) {
    const computedStyle = flattenStyle(style[i]);
    if (computedStyle) for (const key in computedStyle) hasOwnProperty.call(computedStyle, key) && (result[key] = computedStyle[key]);
  }
  return result;
}
const stylesFromProps = /* @__PURE__ */new WeakMap(),
  createDOMProps = (elementType, props, options) => {
    props || (props = emptyObject);
    const {
      accessibilityActiveDescendant,
      accessibilityAtomic,
      accessibilityAutoComplete,
      accessibilityBusy,
      accessibilityChecked,
      accessibilityColumnCount,
      accessibilityColumnIndex,
      accessibilityColumnSpan,
      accessibilityControls,
      accessibilityCurrent,
      accessibilityDescribedBy,
      accessibilityDetails,
      accessibilityDisabled,
      accessibilityErrorMessage,
      accessibilityExpanded,
      accessibilityFlowTo,
      accessibilityHasPopup,
      accessibilityHidden,
      accessibilityInvalid,
      accessibilityKeyShortcuts,
      accessibilityLabel,
      accessibilityLabelledBy,
      accessibilityLevel,
      accessibilityLiveRegion,
      accessibilityModal,
      accessibilityMultiline,
      accessibilityMultiSelectable,
      accessibilityOrientation,
      accessibilityOwns,
      accessibilityPlaceholder,
      accessibilityPosInSet,
      accessibilityPressed,
      accessibilityReadOnly,
      accessibilityRequired,
      /* eslint-disable */
      accessibilityRole,
      /* eslint-enable */
      accessibilityRoleDescription,
      accessibilityRowCount,
      accessibilityRowIndex,
      accessibilityRowSpan,
      accessibilitySelected,
      accessibilitySetSize,
      accessibilitySort,
      accessibilityValueMax,
      accessibilityValueMin,
      accessibilityValueNow,
      accessibilityValueText,
      dataSet,
      focusable,
      nativeID,
      pointerEvents,
      style,
      testID,
      id,
      // Rest
      ...domProps
    } = props;
    for (const key in domProps) reactNativeOnlyProps[key] && delete domProps[key];
    const disabled = accessibilityDisabled,
      role = AccessibilityUtil.propsToAriaRole(props);
    accessibilityActiveDescendant != null && (domProps["aria-activedescendant"] = accessibilityActiveDescendant), accessibilityAtomic != null && (domProps["aria-atomic"] = accessibilityAtomic), accessibilityAutoComplete != null && (domProps["aria-autocomplete"] = accessibilityAutoComplete), accessibilityBusy != null && (domProps["aria-busy"] = accessibilityBusy), accessibilityChecked != null && (domProps["aria-checked"] = accessibilityChecked), accessibilityColumnCount != null && (domProps["aria-colcount"] = accessibilityColumnCount), accessibilityColumnIndex != null && (domProps["aria-colindex"] = accessibilityColumnIndex), accessibilityColumnSpan != null && (domProps["aria-colspan"] = accessibilityColumnSpan), accessibilityControls != null && (domProps["aria-controls"] = processIDRefList(accessibilityControls)), accessibilityCurrent != null && (domProps["aria-current"] = accessibilityCurrent), accessibilityDescribedBy != null && (domProps["aria-describedby"] = processIDRefList(accessibilityDescribedBy)), accessibilityDetails != null && (domProps["aria-details"] = accessibilityDetails), disabled === !0 && (domProps["aria-disabled"] = !0, (elementType === "button" || elementType === "form" || elementType === "input" || elementType === "select" || elementType === "textarea") && (domProps.disabled = !0)), accessibilityErrorMessage != null && (domProps["aria-errormessage"] = accessibilityErrorMessage), accessibilityExpanded != null && (domProps["aria-expanded"] = accessibilityExpanded), accessibilityFlowTo != null && (domProps["aria-flowto"] = processIDRefList(accessibilityFlowTo)), accessibilityHasPopup != null && (domProps["aria-haspopup"] = accessibilityHasPopup), accessibilityHidden === !0 && (domProps["aria-hidden"] = accessibilityHidden), accessibilityInvalid != null && (domProps["aria-invalid"] = accessibilityInvalid), accessibilityKeyShortcuts != null && Array.isArray(accessibilityKeyShortcuts) && (domProps["aria-keyshortcuts"] = accessibilityKeyShortcuts.join(" ")), accessibilityLabel != null && (domProps["aria-label"] = accessibilityLabel), accessibilityLabelledBy != null && (domProps["aria-labelledby"] = processIDRefList(accessibilityLabelledBy)), accessibilityLevel != null && (domProps["aria-level"] = accessibilityLevel), accessibilityLiveRegion != null && (domProps["aria-live"] = accessibilityLiveRegion === "none" ? "off" : accessibilityLiveRegion), accessibilityModal != null && (domProps["aria-modal"] = accessibilityModal), accessibilityMultiline != null && (domProps["aria-multiline"] = accessibilityMultiline), accessibilityMultiSelectable != null && (domProps["aria-multiselectable"] = accessibilityMultiSelectable), accessibilityOrientation != null && (domProps["aria-orientation"] = accessibilityOrientation), accessibilityOwns != null && (domProps["aria-owns"] = processIDRefList(accessibilityOwns)), accessibilityPlaceholder != null && (domProps["aria-placeholder"] = accessibilityPlaceholder), accessibilityPosInSet != null && (domProps["aria-posinset"] = accessibilityPosInSet), accessibilityPressed != null && (domProps["aria-pressed"] = accessibilityPressed), accessibilityReadOnly != null && (domProps["aria-readonly"] = accessibilityReadOnly, (elementType === "input" || elementType === "select" || elementType === "textarea") && (domProps.readOnly = !0)), accessibilityRequired != null && (domProps["aria-required"] = accessibilityRequired, (elementType === "input" || elementType === "select" || elementType === "textarea") && (domProps.required = !0)), role != null && (domProps.role = role === "none" ? "presentation" : role), accessibilityRoleDescription != null && (domProps["aria-roledescription"] = accessibilityRoleDescription), accessibilityRowCount != null && (domProps["aria-rowcount"] = accessibilityRowCount), accessibilityRowIndex != null && (domProps["aria-rowindex"] = accessibilityRowIndex), accessibilityRowSpan != null && (domProps["aria-rowspan"] = accessibilityRowSpan), accessibilitySelected != null && (domProps["aria-selected"] = accessibilitySelected), accessibilitySetSize != null && (domProps["aria-setsize"] = accessibilitySetSize), accessibilitySort != null && (domProps["aria-sort"] = accessibilitySort), accessibilityValueMax != null && (domProps["aria-valuemax"] = accessibilityValueMax), accessibilityValueMin != null && (domProps["aria-valuemin"] = accessibilityValueMin), accessibilityValueNow != null && (domProps["aria-valuenow"] = accessibilityValueNow), accessibilityValueText != null && (domProps["aria-valuetext"] = accessibilityValueText);
    const tmgCN = dataSet ? dataSet.className : void 0,
      tmgID = dataSet ? dataSet.id : void 0;
    if (dataSet != null) {
      for (const dataProp in dataSet) if (!(dataProp === "className" || dataProp === "id") && hasOwnProperty.call(dataSet, dataProp)) {
        const dataName = hyphenateString(dataProp),
          dataValue = dataSet[dataProp];
        dataValue != null && (domProps[`data-${dataName}`] = dataValue);
      }
    }
    focusable === !1 && (domProps.tabIndex = "-1"),
    // These native elements are keyboard focusable by default
    elementType === "a" || elementType === "button" || elementType === "input" || elementType === "select" || elementType === "textarea" ? (focusable === !1 || accessibilityDisabled === !0) && (domProps.tabIndex = "-1") : /* These roles are made keyboard focusable by default */role === "button" || role === "checkbox" || role === "link" || role === "radio" || role === "textbox" || role === "switch" ? focusable !== !1 && (domProps.tabIndex = "0") : focusable === !0 && (domProps.tabIndex = "0");
    const flat = flattenStyle(style);
    let className = tmgCN || "";
    props.className && (className += ` ${props.className}`);
    const stylesAtomic = flat ? getCSSStylesAtomic(flat) : [];
    stylesFromProps.set(domProps, stylesAtomic), domProps.style = stylesAtomic.reduce((acc, [key, value]) => key[0] === "_" || key.startsWith("is_") || key.startsWith("font_") ? (className += ` ${key}`, acc) : (key === "$$css" || key === "" || (acc[key] = value), acc), {}), className && (domProps.className = className);
    const _id = tmgID || id || nativeID;
    return _id && (domProps.id = _id), testID != null && (domProps["data-testid"] = testID), domProps;
  };
export { createDOMProps, stylesFromProps };
