// @tamagui/input@2.0.0-rc.36/dist/esm/v1/Input.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import React from "react";
import { styled } from "@tamagui/core";
import { registerFocusable } from "@tamagui/focusable";
import { useNativeInputRef } from "@tamagui/element";
import { TextInput } from "react-native";
import { styledBody } from "../shared.native.js";
var StyledInput = styled(TextInput, styledBody[0], styledBody[1]),
  Input = StyledInput.styleable(function (inProps, forwardedRef) {
    var {
        // some of destructed props are just to avoid passing them to ...rest because they are not in native.
        type,
        //@ts-ignore
        dirname,
        max,
        min,
        minLength,
        multiple,
        name,
        required,
        step,
        disabled,
        id,
        // @ts-ignore
        caretColor,
        onChange,
        onInput,
        rows,
        enterKeyHint,
        returnKeyType,
        onKeyDown,
        inputMode,
        render,
        ...rest
      } = inProps,
      {
        ref,
        composedRef
      } = useNativeInputRef(forwardedRef),
      secureTextEntry = !1,
      cursorColor = caretColor,
      _returnKeyType = returnKeyType,
      _enterKeyHint = enterKeyHint;
    enterKeyHint === "go" && (_returnKeyType = "go", _enterKeyHint = void 0);
    var _inputMode = inputMode;
    type === "email" ? _inputMode = "email" : type === "tel" ? _inputMode = "tel" : type === "search" ? _inputMode = "search" : type === "url" ? _inputMode = "url" : type === "password" ? (secureTextEntry = !0, _inputMode = "text") : type === "number" ? _inputMode = "numeric" : _inputMode = "text";
    var showSoftInputOnFocus = !0;
    inputMode === "none" && (showSoftInputOnFocus = !1);
    var finalProps = {
      ...rest,
      inputMode: _inputMode,
      showSoftInputOnFocus,
      disabled,
      id,
      cursorColor,
      enterKeyHint: _enterKeyHint,
      returnKeyType: _returnKeyType,
      secureTextEntry,
      numberOfLines: rows || rest.numberOfLines
    };
    return render === "textarea" && (finalProps.multiline = !0), onKeyDown && (finalProps.onKeyPress = function (e) {
      var {
        key
      } = e.nativeEvent;
      (key === "Backspace" || render === "textarea" && key === "Enter" || key.length === 1) && onKeyDown({
        key,
        type: "keydown"
      });
    }, finalProps.onSubmitEditing = function (e) {
      onKeyDown({
        key: "Enter",
        type: "keydown"
      });
    }), (onChange || onInput) && (finalProps.onChange = function (e) {
      var {
        text
      } = e.nativeEvent;
      onChange && onChange({
        target: {
          value: text
        },
        type: "change"
      }), onInput?.({
        target: {
          value: text
        },
        type: "input"
      });
    }), React.useEffect(function () {
      if (id && !disabled) return registerFocusable(id, {
        focusAndSelect: function () {
          var _ref_current;
          (_ref_current = ref.current) === null || _ref_current === void 0 || _ref_current.focus();
        },
        focus: function () {}
      });
    }, [id, disabled]), /* @__PURE__ */_jsx(StyledInput, {
      onChange: function (e) {},
      ref: composedRef,
      ...finalProps
    });
  });
export { Input };
