// @tamagui/input@2.0.0-rc.36/dist/cjs/v1/Input.native.js
"use strict";

var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf,
  __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
    value: mod,
    enumerable: !0
  }) : target, mod)),
  __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
    value: !0
  }), mod);
var Input_native_exports = {};
__export(Input_native_exports, {
  Input: () => Input
});
module.exports = __toCommonJS(Input_native_exports);
var import_jsx_runtime = require("react/jsx-runtime"),
  import_react = __toESM(require("react"), 1),
  import_core = require("@tamagui/core"),
  import_focusable = require("@tamagui/focusable"),
  import_element = require("@tamagui/element"),
  import_react_native = require("react-native"),
  import_shared = require("../shared.native.js"),
  StyledInput = (0, import_core.styled)(import_react_native.TextInput, import_shared.styledBody[0], import_shared.styledBody[1]),
  Input = StyledInput.styleable(function (inProps, forwardedRef) {
    var {
        // some of destructed props are just to avoid passing them to ...rest because they are not in native.
        type,
        //@ts-ignore
        dirname,
        max,
        min,
        minLength,
        multiple,
        name,
        required,
        step,
        disabled,
        id,
        // @ts-ignore
        caretColor,
        onChange,
        onInput,
        rows,
        enterKeyHint,
        returnKeyType,
        onKeyDown,
        inputMode,
        render,
        ...rest
      } = inProps,
      {
        ref,
        composedRef
      } = (0, import_element.useNativeInputRef)(forwardedRef),
      secureTextEntry = !1,
      cursorColor = caretColor,
      _returnKeyType = returnKeyType,
      _enterKeyHint = enterKeyHint;
    enterKeyHint === "go" && (_returnKeyType = "go", _enterKeyHint = void 0);
    var _inputMode = inputMode;
    type === "email" ? _inputMode = "email" : type === "tel" ? _inputMode = "tel" : type === "search" ? _inputMode = "search" : type === "url" ? _inputMode = "url" : type === "password" ? (secureTextEntry = !0, _inputMode = "text") : type === "number" ? _inputMode = "numeric" : _inputMode = "text";
    var showSoftInputOnFocus = !0;
    inputMode === "none" && (showSoftInputOnFocus = !1);
    var finalProps = {
      ...rest,
      inputMode: _inputMode,
      showSoftInputOnFocus,
      disabled,
      id,
      cursorColor,
      enterKeyHint: _enterKeyHint,
      returnKeyType: _returnKeyType,
      secureTextEntry,
      numberOfLines: rows || rest.numberOfLines
    };
    return render === "textarea" && (finalProps.multiline = !0), onKeyDown && (finalProps.onKeyPress = function (e) {
      var {
        key
      } = e.nativeEvent;
      (key === "Backspace" || render === "textarea" && key === "Enter" || key.length === 1) && onKeyDown({
        key,
        type: "keydown"
      });
    }, finalProps.onSubmitEditing = function (e) {
      onKeyDown({
        key: "Enter",
        type: "keydown"
      });
    }), (onChange || onInput) && (finalProps.onChange = function (e) {
      var {
        text
      } = e.nativeEvent;
      onChange && onChange({
        target: {
          value: text
        },
        type: "change"
      }), onInput?.({
        target: {
          value: text
        },
        type: "input"
      });
    }), import_react.default.useEffect(function () {
      if (id && !disabled) return (0, import_focusable.registerFocusable)(id, {
        focusAndSelect: function () {
          var _ref_current;
          (_ref_current = ref.current) === null || _ref_current === void 0 || _ref_current.focus();
        },
        focus: function () {}
      });
    }, [id, disabled]), /* @__PURE__ */(0, import_jsx_runtime.jsx)(StyledInput, {
      onChange: function (e) {},
      ref: composedRef,
      ...finalProps
    });
  });
