// @tamagui/vite-plugin@2.0.0-rc.36/dist/esm/loadTamagui.mjs
import * as StaticWorker from "@tamagui/static-worker";
const LOAD_STATE_KEY = "__tamagui_load_state__";
function getLoadState() {
  return globalThis[LOAD_STATE_KEY] || (globalThis[LOAD_STATE_KEY] = {
    loadPromise: null,
    loadedOptions: null,
    fullConfigLoaded: !1,
    fullConfigLoadPromise: null
  }), globalThis[LOAD_STATE_KEY];
}
function getTamaguiOptions() {
  return getLoadState().loadedOptions;
}
function getLoadPromise() {
  return getLoadState().loadPromise;
}
async function loadTamaguiBuildConfig(optionsIn) {
  const state = getLoadState();
  return state.loadedOptions ? state.loadedOptions : (state.loadPromise || (state.loadPromise = (async () => {
    const options = await StaticWorker.loadTamaguiBuildConfig({
      ...optionsIn,
      platform: "web"
    });
    return state.loadedOptions = options, options;
  })()), state.loadPromise);
}
async function ensureFullConfigLoaded() {
  const state = getLoadState();
  if (!state.fullConfigLoaded) return state.fullConfigLoadPromise || (state.fullConfigLoadPromise = (async () => {
    const options = await loadTamaguiBuildConfig();
    !options.disableWatchTamaguiConfig && !options.disable && (await StaticWorker.loadTamagui({
      components: ["tamagui"],
      platform: "web",
      ...options
    })), state.fullConfigLoaded = !0;
  })()), state.fullConfigLoadPromise;
}
async function cleanup() {
  await StaticWorker.destroyPool();
  const state = getLoadState();
  state.loadPromise = null, state.loadedOptions = null, state.fullConfigLoaded = !1, state.fullConfigLoadPromise = null;
}
export { cleanup, ensureFullConfigLoaded, getLoadPromise, getTamaguiOptions, loadTamaguiBuildConfig };
