// @tamagui/vite-plugin@2.0.0-rc.36/dist/esm/plugin.mjs
import * as Static from "@tamagui/static-worker";
import { getPragmaOptions } from "@tamagui/static-worker";
import { createHash } from "node:crypto";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { normalizePath, transformWithEsbuild } from "vite";
import { loadTamaguiBuildConfig, getLoadPromise, getTamaguiOptions, ensureFullConfigLoaded } from "./loadTamagui.mjs";
const resolve = name => fileURLToPath(import.meta.resolve(name)),
  CACHE_KEY = "__tamagui_vite_cache__",
  CACHE_SIZE_KEY = "__tamagui_vite_cache_size__",
  PENDING_KEY = "__tamagui_vite_pending__";
function getSharedCache() {
  return globalThis[CACHE_KEY] || (globalThis[CACHE_KEY] = {}), globalThis[CACHE_KEY];
}
function getSharedCacheSize() {
  return globalThis[CACHE_SIZE_KEY] || 0;
}
function setSharedCacheSize(size) {
  globalThis[CACHE_SIZE_KEY] = size;
}
function clearSharedCache() {
  globalThis[CACHE_KEY] = {}, globalThis[CACHE_SIZE_KEY] = 0;
}
function getPendingExtractions() {
  return globalThis[PENDING_KEY] || (globalThis[PENDING_KEY] = /* @__PURE__ */new Map()), globalThis[PENDING_KEY];
}
function tamaguiAliases(options = {}) {
  const aliases = [];
  if (options.svg && aliases.push({
    find: "react-native-svg",
    replacement: resolve("@tamagui/react-native-svg")
  }), options.rnwLite) {
    const rnwl = resolve(options.rnwLite === "without-animated" ? "@tamagui/react-native-web-lite/without-animated" : "@tamagui/react-native-web-lite"),
      rnwlBase = path.dirname(resolve("@tamagui/react-native-web-lite/package.json"));
    aliases.push({
      // map deep RNW paths like dist/exports/StyleSheet/preprocess to rnw-lite's flat structure
      // extracts the final path segment (e.g. "preprocess" or "createReactDOMStyle")
      find: /^react-native(?:-web)?\/dist\/(?:exports|modules)\/.*\/([^/]+)$/,
      replacement: `${rnwlBase}/dist/esm/$1.mjs`
    }, {
      find: /^react-native$/,
      replacement: rnwl
    }, {
      find: /^react-native\/(Libraries\/Utilities\/codegenNativeComponent|Libraries\/Utilities\/codegenNativeCommand)$/,
      replacement: `${rnwlBase}/$1`
    }, {
      find: "react-native/package.json",
      replacement: resolve("@tamagui/react-native-web-lite/package.json")
    }, {
      find: /^react-native-web$/,
      replacement: rnwl
    });
  }
  return aliases;
}
function tamaguiPlugin({
  disableResolveConfig,
  ...tamaguiOptionsIn
} = {}) {
  let shouldExtract = !tamaguiOptionsIn.disableExtraction,
    watcher;
  const enableNativeEnv = !!globalThis.__vxrnEnableNativeEnv,
    extensions = [".web.mjs", ".web.js", ".web.jsx", ".web.ts", ".web.tsx", ".mjs", ".js", ".mts", ".ts", ".jsx", ".tsx", ".json"];
  loadTamaguiBuildConfig(tamaguiOptionsIn);
  const ensureLoaded = async () => {
      const promise = getLoadPromise();
      promise && (await promise);
      const options = getTamaguiOptions();
      return options && (shouldExtract = !options.disableExtraction), options;
    },
    getHash = input => createHash("sha1").update(input).digest("base64"),
    memoryCache = getSharedCache(),
    cssMap = /* @__PURE__ */new Map();
  let config, server;
  const virtualExt = ".tamagui.css",
    getAbsoluteVirtualFileId = filePath => filePath.startsWith(config.root) ? filePath : normalizePath(path.join(config.root, filePath));
  function isNotClient(environment) {
    return environment?.name && environment.name !== "client";
  }
  function isNative(environment) {
    return environment?.name && (environment.name === "ios" || environment.name === "android");
  }
  function invalidateModule(absoluteId) {
    if (!server) return;
    const {
        moduleGraph
      } = server,
      modules = moduleGraph.getModulesByFile(absoluteId);
    if (modules) for (const module of modules) moduleGraph.invalidateModule(module), module.lastHMRTimestamp = module.lastInvalidationTimestamp || Date.now();
  }
  return [{
    name: "tamagui",
    enforce: "pre",
    configureServer(_server) {
      server = _server;
    },
    async buildEnd() {
      await watcher?.then(res => {
        res?.dispose();
      });
    },
    async transform(code, id) {
      if (id.includes("expo-linear-gradient")) return transformWithEsbuild(code, id, {
        loader: "jsx",
        jsx: "automatic"
      });
    },
    async config(_, env) {
      const options = await ensureLoaded();
      if (!options) throw new Error("No tamagui options loaded");
      return options.disableWatchTamaguiConfig || (watcher = Static.watchTamaguiConfig({
        components: ["tamagui"],
        config: "./src/tamagui.config.ts",
        ...options
      }).catch(err => {
        console.error(` [Tamagui] Error watching config: ${err}`);
      })), {
        envPrefix: ["TAMAGUI_"],
        environments: {
          client: {
            define: {
              "process.env.TAMAGUI_IS_CLIENT": JSON.stringify(!0),
              "process.env.TAMAGUI_ENVIRONMENT": '"client"'
            }
          }
        },
        define: {
          // reanimated support
          _frameTimestamp: void 0,
          _WORKLET: !1,
          __DEV__: `${env.mode === "development"}`,
          "process.env.NODE_ENV": JSON.stringify(process.env.NODE_ENV || env.mode),
          "process.env.ENABLE_RSC": JSON.stringify(process.env.ENABLE_RSC || ""),
          "process.env.ENABLE_STEPS": JSON.stringify(process.env.ENABLE_STEPS || ""),
          "process.env.IS_STATIC": JSON.stringify(!1),
          ...(env.mode === "production" && {
            "process.env.TAMAGUI_OPTIMIZE_THEMES": JSON.stringify(!0)
          })
        },
        resolve: disableResolveConfig || enableNativeEnv ? {} : {
          extensions,
          alias: {
            ...(options.platform !== "native" && {
              "react-native/Libraries/Renderer/shims/ReactFabric": resolve("@tamagui/proxy-worm"),
              "react-native/Libraries/Utilities/codegenNativeComponent": resolve("@tamagui/proxy-worm"),
              "react-native-svg": resolve("@tamagui/react-native-svg"),
              ...(!options?.useReactNativeWebLite && {
                "react-native": resolve("react-native-web")
              })
            })
          }
        }
      };
    }
  }, {
    name: "tamagui-rnw-lite",
    config() {
      if (enableNativeEnv) return {};
      const options = getTamaguiOptions();
      return options?.useReactNativeWebLite ? {
        resolve: {
          alias: tamaguiAliases({
            rnwLite: options.useReactNativeWebLite
          })
        },
        optimizeDeps: {
          // upstream react-native-web must not be pre-bundled when aliased to lite
          exclude: ["react-native-web"]
        }
      } : {};
    }
  }, {
    name: "tamagui-extract",
    enforce: "pre",
    async config(userConf) {
      const options = await ensureLoaded();
      userConf.optimizeDeps ||= {}, userConf.optimizeDeps.include ||= [], userConf.optimizeDeps.include.push("inline-style-prefixer"), shouldExtract && userConf.optimizeDeps.include.push("@tamagui/core/inject-styles");
    },
    async configResolved(resolvedConfig) {
      config = resolvedConfig;
    },
    async resolveId(source) {
      if (!shouldExtract || isNative(this.environment) || isNotClient(this.environment)) return;
      const [validId, query] = source.split("?");
      if (!validId.endsWith(virtualExt)) return;
      const absoluteId = source.startsWith(config.root) ? source : getAbsoluteVirtualFileId(validId);
      if (cssMap.has(absoluteId)) return absoluteId + (query ? `?${query}` : "");
    },
    async load(id) {
      if (!shouldExtract || getTamaguiOptions()?.disable || isNative(this.environment) || isNotClient(this.environment)) return;
      const [validId] = id.split("?");
      return cssMap.get(validId);
    },
    transform: {
      order: "pre",
      async handler(code, id) {
        const options = await ensureLoaded();
        if (await ensureFullConfigLoaded(), options?.disable || isNative(this.environment)) return;
        const [validId] = id.split("?");
        if (!validId.endsWith(".tsx")) return;
        const {
          shouldDisable,
          shouldPrintDebug
        } = await getPragmaOptions({
          source: code,
          path: validId
        });
        if (shouldPrintDebug && (console.trace(`Current file: ${id} in environment: ${this.environment?.name}, shouldDisable: ${shouldDisable}`), console.info(`

Original source:
${code}

`)), shouldDisable) return;
        const isSSR = isNotClient(this.environment),
          cacheKey = getHash(`${code}${id}`),
          pending = getPendingExtractions(),
          formatResult = entry => ({
            code: !isSSR && entry.cssImport ? `${entry.js}
${entry.cssImport}` : entry.js,
            map: entry.map
          }),
          cached = memoryCache[cacheKey];
        if (cached) return process.env.DEBUG_TAMAGUI_CACHE && console.info(`[tamagui-cache] HIT ${this.environment?.name || "unknown"} ${id.split("/").pop()} key=${cacheKey.slice(0, 8)}`), formatResult(cached);
        const pendingExtraction = pending.get(cacheKey);
        if (pendingExtraction) {
          process.env.DEBUG_TAMAGUI_CACHE && console.info(`[tamagui-cache] WAIT ${this.environment?.name || "unknown"} ${id.split("/").pop()} key=${cacheKey.slice(0, 8)}`);
          const result = await pendingExtraction;
          return result ? formatResult(result) : void 0;
        }
        process.env.DEBUG_TAMAGUI_CACHE && console.info(`[tamagui-cache] EXTRACT ${this.environment?.name || "unknown"} ${id.split("/").pop()} key=${cacheKey.slice(0, 8)}`);
        const extractionPromise = (async () => {
          let extracted;
          try {
            extracted = await Static.extractToClassNames({
              source: code,
              sourcePath: validId,
              options,
              shouldPrintDebug
            });
          } catch (err) {
            return process.env.DEBUG_TAMAGUI_CACHE && console.info(`[tamagui-cache] ERROR extracting ${id.split("/").pop()}:`, err), console.error(err instanceof Error ? err.message : String(err)), null;
          }
          if (!extracted) return process.env.DEBUG_TAMAGUI_CACHE && console.info(`[tamagui-cache] no extraction result for ${id.split("/").pop()}`), null;
          const rootRelativeId = `${validId}${virtualExt}`,
            absoluteId = getAbsoluteVirtualFileId(rootRelativeId);
          let cssImport = null;
          extracted.styles && (this.addWatchFile(rootRelativeId), server && cssMap.has(absoluteId) && invalidateModule(rootRelativeId), cssImport = `import "${rootRelativeId}";`, cssMap.set(absoluteId, extracted.styles));
          const jsCode = extracted.js.toString(),
            cacheEntry = {
              js: jsCode,
              map: extracted.map,
              cssImport
            },
            newSize = getSharedCacheSize() + jsCode.length;
          return newSize > 67108864 ? clearSharedCache() : setSharedCacheSize(newSize), memoryCache[cacheKey] = cacheEntry, process.env.DEBUG_TAMAGUI_CACHE && console.info(`[tamagui-cache] WRITE key=${cacheKey.slice(0, 8)} cacheSize=${Object.keys(memoryCache).length}`), cacheEntry;
        })();
        pending.set(cacheKey, extractionPromise);
        try {
          const result = await extractionPromise;
          return result ? formatResult(result) : void 0;
        } finally {
          pending.delete(cacheKey);
        }
      }
    }
  }];
}
export { tamaguiAliases, tamaguiPlugin };
