// @tamagui/web@2.0.0-rc.36/dist/cjs/helpers/webPropsToSkip.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var webPropsToSkip_native_exports = {};
__export(webPropsToSkip_native_exports, {
  webPropsToSkip: () => webPropsToSkip
});
module.exports = __toCommonJS(webPropsToSkip_native_exports);
var import_helpers = require("@tamagui/helpers"),
  {
    pointerEvents: _,
    ...webOnlyStylePropsViewWithoutPointerEvents
  } = import_helpers.webOnlyStylePropsView,
  webPropsToSkip = {
    // Reuse web-only style props from webOnlyStyleProps (excluding pointerEvents)
    ...webOnlyStylePropsViewWithoutPointerEvents,
    ...import_helpers.webOnlyStylePropsText,
    // Web-only event handlers
    onClick: 1,
    onDoubleClick: 1,
    onContextMenu: 1,
    onMouseEnter: 1,
    onMouseLeave: 1,
    onMouseMove: 1,
    onMouseOver: 1,
    onMouseOut: 1,
    onMouseDown: 1,
    onMouseUp: 1,
    onWheel: 1,
    onKeyDown: 1,
    onKeyUp: 1,
    onKeyPress: 1,
    onPointerDown: 1,
    onPointerMove: 1,
    onPointerUp: 1,
    onPointerCancel: 1,
    onPointerEnter: 1,
    onPointerLeave: 1,
    onDrag: 1,
    onDragStart: 1,
    onDragEnd: 1,
    onDragEnter: 1,
    onDragLeave: 1,
    onDragOver: 1,
    onDrop: 1,
    onChange: 1,
    onInput: 1,
    onBeforeInput: 1,
    onScroll: 1,
    onCopy: 1,
    onCut: 1,
    onPaste: 1,
    // Other web-only props
    htmlFor: 1,
    dangerouslySetInnerHTML: 1
  };
