// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/amberA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var amberA_exports = {};
__export(amberA_exports, {
  amberA: () => amberA
});
module.exports = __toCommonJS(amberA_exports);
const amberA = {
  amberA1: "#e63c0006",
  amberA2: "#fd9b000d",
  amberA3: "#fa820022",
  amberA4: "#fc820032",
  amberA5: "#fd8b0041",
  amberA6: "#fd9b0051",
  amberA7: "#ffab2567",
  amberA8: "#ffae3587",
  amberA9: "#ffc53d",
  amberA10: "#ffd60a",
  amberA11: "#ffca16",
  amberA12: "#ffe7b3"
};
