// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/blueA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var blueA_exports = {};
__export(blueA_exports, {
  blueA: () => blueA
});
module.exports = __toCommonJS(blueA_exports);
const blueA = {
  blueA1: "#004df211",
  blueA2: "#1166fb18",
  blueA3: "#0077ff3a",
  blueA4: "#0075ff57",
  blueA5: "#0081fd6b",
  blueA6: "#0f89fd7f",
  blueA7: "#2a91fe98",
  blueA8: "#3094feb9",
  blueA9: "#0090ff",
  blueA10: "#3b9eff",
  blueA11: "#70b8ff",
  blueA12: "#c2e6ff"
};
