// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/brown.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var brown_exports = {};
__export(brown_exports, {
  brown: () => brown
});
module.exports = __toCommonJS(brown_exports);
const brown = {
  brown1: "#12110f",
  brown2: "#1c1816",
  brown3: "#28211d",
  brown4: "#322922",
  brown5: "#3e3128",
  brown6: "#4d3c2f",
  brown7: "#614a39",
  brown8: "#7c5f46",
  brown9: "#ad7f58",
  brown10: "#b88c67",
  brown11: "#dbb594",
  brown12: "#f2e1ca"
};
