// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/goldA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var goldA_exports = {};
__export(goldA_exports, {
  goldA: () => goldA
});
module.exports = __toCommonJS(goldA_exports);
var goldA = {
  goldA1: "#91911102",
  goldA2: "#f9e29d0b",
  goldA3: "#f8ecbb15",
  goldA4: "#ffeec41e",
  goldA5: "#feecc22a",
  goldA6: "#feebcb37",
  goldA7: "#ffedcd48",
  goldA8: "#fdeaca5f",
  goldA9: "#ffdba690",
  goldA10: "#fedfb09d",
  goldA11: "#fee7c6c8",
  goldA12: "#fef7ede7"
};
