// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/grayA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var grayA_exports = {};
__export(grayA_exports, {
  grayA: () => grayA
});
module.exports = __toCommonJS(grayA_exports);
var grayA = {
  grayA1: "#00000000",
  grayA2: "#ffffff09",
  grayA3: "#ffffff12",
  grayA4: "#ffffff1b",
  grayA5: "#ffffff22",
  grayA6: "#ffffff2c",
  grayA7: "#ffffff3b",
  grayA8: "#ffffff55",
  grayA9: "#ffffff64",
  grayA10: "#ffffff72",
  grayA11: "#ffffffaf",
  grayA12: "#ffffffed"
};
