// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/indigoA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var indigoA_exports = {};
__export(indigoA_exports, {
  indigoA: () => indigoA
});
module.exports = __toCommonJS(indigoA_exports);
var indigoA = {
  indigoA1: "#1133ff0f",
  indigoA2: "#3354fa17",
  indigoA3: "#2f62ff3c",
  indigoA4: "#3566ff57",
  indigoA5: "#4171fd6b",
  indigoA6: "#5178fd7c",
  indigoA7: "#5a7fff90",
  indigoA8: "#5b81feac",
  indigoA9: "#4671ffdb",
  indigoA10: "#5c7efee3",
  indigoA11: "#9eb1ff",
  indigoA12: "#d6e1ff"
};
