// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/mauveA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var mauveA_exports = {};
__export(mauveA_exports, {
  mauveA: () => mauveA
});
module.exports = __toCommonJS(mauveA_exports);
var mauveA = {
  mauveA1: "#00000000",
  mauveA2: "#f5f4f609",
  mauveA3: "#ebeaf814",
  mauveA4: "#eee5f81d",
  mauveA5: "#efe6fe25",
  mauveA6: "#f1e6fd30",
  mauveA7: "#eee9ff40",
  mauveA8: "#eee7ff5d",
  mauveA9: "#eae6fd6e",
  mauveA10: "#ece9fd7c",
  mauveA11: "#f5f1ffb7",
  mauveA12: "#fdfdffef"
};
