// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/mintA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var mintA_exports = {};
__export(mintA_exports, {
  mintA: () => mintA
});
module.exports = __toCommonJS(mintA_exports);
var mintA = {
  mintA1: "#00dede05",
  mintA2: "#00f9f90b",
  mintA3: "#00fff61d",
  mintA4: "#00fff42c",
  mintA5: "#00fff23a",
  mintA6: "#0effeb4a",
  mintA7: "#34fde55e",
  mintA8: "#41ffdf76",
  mintA9: "#92ffe7e9",
  mintA10: "#aefeedf5",
  mintA11: "#67ffded2",
  mintA12: "#cbfee9f5"
};
