// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/mint.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var mint_exports = {};
__export(mint_exports, {
  mint: () => mint
});
module.exports = __toCommonJS(mint_exports);
var mint = {
  mint1: "#0e1515",
  mint2: "#0f1b1b",
  mint3: "#092c2b",
  mint4: "#003a38",
  mint5: "#004744",
  mint6: "#105650",
  mint7: "#1e685f",
  mint8: "#277f70",
  mint9: "#86ead4",
  mint10: "#a8f5e5",
  mint11: "#58d5ba",
  mint12: "#c4f5e1"
};
