// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/pinkA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var pinkA_exports = {};
__export(pinkA_exports, {
  pinkA: () => pinkA
});
module.exports = __toCommonJS(pinkA_exports);
var pinkA = {
  pinkA1: "#f412bc09",
  pinkA2: "#f420bb12",
  pinkA3: "#fe37cc29",
  pinkA4: "#fc1ec43f",
  pinkA5: "#fd35c24e",
  pinkA6: "#fd51c75f",
  pinkA7: "#fd62c87b",
  pinkA8: "#ff68c8a2",
  pinkA9: "#fe49bcd4",
  pinkA10: "#ff5cc0dc",
  pinkA11: "#ff8dcc",
  pinkA12: "#ffd3ecfd"
};
