// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/purpleA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var purpleA_exports = {};
__export(purpleA_exports, {
  purpleA: () => purpleA
});
module.exports = __toCommonJS(purpleA_exports);
const purpleA = {
  purpleA1: "#b412f90b",
  purpleA2: "#b744f714",
  purpleA3: "#c150ff2d",
  purpleA4: "#bb53fd42",
  purpleA5: "#be5cfd51",
  purpleA6: "#c16dfd61",
  purpleA7: "#c378fd7a",
  purpleA8: "#c47effa4",
  purpleA9: "#b661ffc2",
  purpleA10: "#bc6fffcd",
  purpleA11: "#d19dff",
  purpleA12: "#f1ddfffa"
};
