// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/redA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var redA_exports = {};
__export(redA_exports, {
  redA: () => redA
});
module.exports = __toCommonJS(redA_exports);
const redA = {
  redA1: "#f4121209",
  redA2: "#f22f3e11",
  redA3: "#ff173f2d",
  redA4: "#fe0a3b44",
  redA5: "#ff204756",
  redA6: "#ff3e5668",
  redA7: "#ff536184",
  redA8: "#ff5d61b0",
  redA9: "#fe4e54e4",
  redA10: "#ff6465eb",
  redA11: "#ff9592",
  redA12: "#ffd1d9"
};
