// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/sageA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var sageA_exports = {};
__export(sageA_exports, {
  sageA: () => sageA
});
module.exports = __toCommonJS(sageA_exports);
var sageA = {
  sageA1: "#00000000",
  sageA2: "#f0f2f108",
  sageA3: "#f3f5f412",
  sageA4: "#f2fefd1a",
  sageA5: "#f1fbfa22",
  sageA6: "#edfbf42d",
  sageA7: "#edfcf73c",
  sageA8: "#ebfdf657",
  sageA9: "#dffdf266",
  sageA10: "#e5fdf674",
  sageA11: "#f4fefbb0",
  sageA12: "#fdfffeed"
};
