// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/sandA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var sandA_exports = {};
__export(sandA_exports, {
  sandA: () => sandA
});
module.exports = __toCommonJS(sandA_exports);
var sandA = {
  sandA1: "#00000000",
  sandA2: "#f4f4f309",
  sandA3: "#f6f6f513",
  sandA4: "#fefef31b",
  sandA5: "#fbfbeb23",
  sandA6: "#fffaed2d",
  sandA7: "#fffbed3c",
  sandA8: "#fff9eb57",
  sandA9: "#fffae965",
  sandA10: "#fffdee73",
  sandA11: "#fffcf4b0",
  sandA12: "#fffffded"
};
