// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/skyA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var skyA_exports = {};
__export(skyA_exports, {
  skyA: () => skyA
});
module.exports = __toCommonJS(skyA_exports);
const skyA = {
  skyA1: "#0044ff0f",
  skyA2: "#1171fb18",
  skyA3: "#1184fc33",
  skyA4: "#128fff49",
  skyA5: "#1c9dfd5d",
  skyA6: "#28a5ff72",
  skyA7: "#2badfe8b",
  skyA8: "#1db2fea9",
  skyA9: "#7ce3fffe",
  skyA10: "#a8eeff",
  skyA11: "#7cd3ffef",
  skyA12: "#c2f3ff"
};
