// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/tealA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var tealA_exports = {};
__export(tealA_exports, {
  tealA: () => tealA
});
module.exports = __toCommonJS(tealA_exports);
const tealA = {
  tealA1: "#00deab05",
  tealA2: "#12fbe60c",
  tealA3: "#00ffe61e",
  tealA4: "#00ffe92d",
  tealA5: "#00ffea3b",
  tealA6: "#1cffe84b",
  tealA7: "#2efde85f",
  tealA8: "#32ffe775",
  tealA9: "#13ffe49f",
  tealA10: "#0dffe0ae",
  tealA11: "#0afed5d6",
  tealA12: "#b8ffebef"
};
