// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/tomatoA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var tomatoA_exports = {};
__export(tomatoA_exports, {
  tomatoA: () => tomatoA
});
module.exports = __toCommonJS(tomatoA_exports);
const tomatoA = {
  tomatoA1: "#f1121208",
  tomatoA2: "#ff55330f",
  tomatoA3: "#ff35232b",
  tomatoA4: "#fd201142",
  tomatoA5: "#fe332153",
  tomatoA6: "#ff4f3864",
  tomatoA7: "#fd644a7d",
  tomatoA8: "#fe6d4ea7",
  tomatoA9: "#fe5431e4",
  tomatoA10: "#ff6847eb",
  tomatoA11: "#ff977d",
  tomatoA12: "#ffd6cefb"
};
