// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/violetA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var violetA_exports = {};
__export(violetA_exports, {
  violetA: () => violetA
});
module.exports = __toCommonJS(violetA_exports);
const violetA = {
  violetA1: "#4422ff0f",
  violetA2: "#853ff916",
  violetA3: "#8354fe36",
  violetA4: "#7d51fd50",
  violetA5: "#845ffd5f",
  violetA6: "#8f6cfd6d",
  violetA7: "#9879ff83",
  violetA8: "#977dfea8",
  violetA9: "#8668ffcc",
  violetA10: "#9176fed7",
  violetA11: "#baa7ff",
  violetA12: "#e3defffe"
};
