// @tamagui/colors@2.0.0-rc.36/dist/cjs/dark/yellowA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var yellowA_exports = {};
__export(yellowA_exports, {
  yellowA: () => yellowA
});
module.exports = __toCommonJS(yellowA_exports);
var yellowA = {
  yellowA1: "#d1510004",
  yellowA2: "#f9b4000b",
  yellowA3: "#ffaa001e",
  yellowA4: "#fdb70028",
  yellowA5: "#febb0036",
  yellowA6: "#fec40046",
  yellowA7: "#fdcb225c",
  yellowA8: "#fdca327b",
  yellowA9: "#ffe629",
  yellowA10: "#ffff57",
  yellowA11: "#fee949f5",
  yellowA12: "#fef6baf6"
};
