// @tamagui/static-worker@2.0.0-rc.36/dist/esm/index.mjs
import { fileURLToPath } from "node:url";
import Piscina from "piscina";
const getPragmaOptions = async props => {
    const {
      default: Static
    } = await import("@tamagui/static");
    return Static.getPragmaOptions(props);
  },
  getWorkerPath = () => typeof import.meta < "u" && import.meta.url ? fileURLToPath(import.meta.resolve("@tamagui/static/worker")).replace(/\.mjs$/, ".js") : require.resolve("@tamagui/static/worker").replace(/\.mjs$/, ".js"),
  POOL_KEY = "__tamagui_piscina_pool__",
  CLOSING_KEY = "__tamagui_piscina_closing__",
  TASK_COUNT_KEY = "__tamagui_piscina_task_count__",
  RECYCLING_KEY = "__tamagui_piscina_recycling__",
  MAX_TASKS_BEFORE_RECYCLE = 1e3;
function getSharedPool() {
  return globalThis[POOL_KEY] ?? null;
}
function setSharedPool(pool) {
  globalThis[POOL_KEY] = pool;
}
function isClosing() {
  return globalThis[CLOSING_KEY] === !0;
}
function setClosing(value) {
  globalThis[CLOSING_KEY] = value;
}
function isRecycling() {
  return globalThis[RECYCLING_KEY] === !0;
}
function setRecycling(value) {
  globalThis[RECYCLING_KEY] = value;
}
function getTaskCount() {
  return globalThis[TASK_COUNT_KEY] ?? 0;
}
function incrementTaskCount() {
  const count = getTaskCount() + 1;
  return globalThis[TASK_COUNT_KEY] = count, count;
}
function resetTaskCount() {
  globalThis[TASK_COUNT_KEY] = 0;
}
function createPool() {
  const pool = new Piscina({
    filename: getWorkerPath(),
    // each worker loads and caches config independently
    minThreads: 2,
    maxThreads: 2,
    // Never terminate due to idle - worker stays alive until close() or process exit
    // This prevents "Terminating worker thread" errors from Piscina during idle
    idleTimeout: Number.POSITIVE_INFINITY
    // no resourceLimits - we rely on task-based recycling instead
    // V8 resourceLimits cause "Terminating worker thread" messages when hit
  });
  return pool.on("error", err => {
    isClosing() || isRecycling() || (err && typeof err == "object" && "message" in err ? String(err.message) : "").includes("Terminating worker thread") || console.error("[tamagui] Worker pool error:", err);
  }), pool;
}
function getPool() {
  let pool = getSharedPool();
  return pool || (pool = createPool(), setSharedPool(pool)), pool;
}
async function loadTamagui(options) {
  const pool = getPool(),
    task = {
      type: "extractToClassNames",
      source: "// dummy",
      sourcePath: "__dummy__.tsx",
      options: {
        components: ["tamagui"],
        ...options
      },
      shouldPrintDebug: !1
    };
  try {
    return await pool.run(task, {
      name: "runTask"
    }), {
      success: !0
    };
  } catch (error) {
    throw console.error("[static-worker] Error loading Tamagui config:", error), error;
  }
}
async function recyclePool(options) {
  if (isClosing() || isRecycling()) return;
  const oldPool = getSharedPool();
  if (!oldPool) return;
  setRecycling(!0);
  const start = Date.now();
  try {
    const originalStderr = process.stderr.write.bind(process.stderr),
      originalStdout = process.stdout.write.bind(process.stdout),
      filter = (chunk, ...args) => !!(typeof chunk == "string" ? chunk : chunk?.toString?.() || "").includes("Terminating worker thread");
    process.stderr.write = (chunk, ...args) => filter(chunk) ? !0 : originalStderr(chunk, ...args), process.stdout.write = (chunk, ...args) => filter(chunk) ? !0 : originalStdout(chunk, ...args);
    const newPool = createPool();
    setSharedPool(newPool);
    const warmupTask = {
      type: "extractToClassNames",
      source: "// warmup",
      sourcePath: "__warmup__.tsx",
      options: {
        ...options,
        // skip the "built config" log on warmup since it's a recycle
        _skipBuildLog: !0
      },
      shouldPrintDebug: !1
    };
    await newPool.run(warmupTask, {
      name: "runTask"
    }), oldPool.removeAllListeners(), oldPool.destroy().catch(() => {}), setTimeout(() => {
      process.stderr.write = originalStderr, process.stdout.write = originalStdout;
    }), console.log(`  \u267B\uFE0F  [tamagui] recycled worker pool (${Date.now() - start}ms)`);
  } finally {
    setRecycling(!1);
  }
}
async function loadTamaguiBuildConfig(tamaguiOptions) {
  const {
    default: Static
  } = await import("@tamagui/static");
  return Static.loadTamaguiBuildConfigAsync(tamaguiOptions);
}
async function extractToClassNames(params) {
  const {
    source,
    sourcePath = "",
    options,
    shouldPrintDebug = !1
  } = params;
  if (typeof source != "string") throw new Error("`source` must be a string of javascript");
  const task = {
      type: "extractToClassNames",
      source,
      sourcePath,
      options,
      shouldPrintDebug
    },
    result = await getPool().run(task, {
      name: "runTask"
    });
  if (!result.success) {
    const errorMessage = [`[tamagui-extract] Error processing file: ${sourcePath || "(unknown)"}`, "", result.error, result.stack ? `
${result.stack}` : ""].filter(Boolean).join(`
`);
    throw new Error(errorMessage);
  }
  return incrementTaskCount() >= MAX_TASKS_BEFORE_RECYCLE && (resetTaskCount(), recyclePool(options).catch(() => {})), result.data;
}
async function extractToNative(sourceFileName, sourceCode, options) {
  const task = {
      type: "extractToNative",
      sourceFileName,
      sourceCode,
      options
    },
    result = await getPool().run(task, {
      name: "runTask"
    });
  if (!result.success) {
    const errorMessage = [`[tamagui-extract] Error processing file: ${sourceFileName || "(unknown)"}`, "", result.error, result.stack ? `
${result.stack}` : ""].filter(Boolean).join(`
`);
    throw new Error(errorMessage);
  }
  return incrementTaskCount() >= MAX_TASKS_BEFORE_RECYCLE && (resetTaskCount(), recyclePool(options).catch(() => {})), result.data;
}
async function watchTamaguiConfig(options) {
  const {
      default: Static
    } = await import("@tamagui/static"),
    watcher = await Static.watchTamaguiConfig(options);
  if (!watcher) return;
  const originalDispose = watcher.dispose;
  return {
    dispose: () => {
      originalDispose(), getSharedPool() && clearWorkerCache();
    }
  };
}
async function clearWorkerCache() {
  const pool = getSharedPool();
  if (!pool || isClosing()) return;
  const task = {
    type: "clearCache"
  };
  await pool.run(task, {
    name: "runTask"
  });
}
async function destroyPool() {
  const pool = getSharedPool();
  if (pool) {
    setClosing(!0);
    try {
      await pool.close();
    } finally {
      setSharedPool(null), setClosing(!1);
    }
  }
}
function getPoolStats() {
  const pool = getSharedPool();
  return pool ? {
    threads: pool.threads.length,
    queueSize: pool.queueSize,
    completed: pool.completed,
    duration: pool.duration,
    utilization: pool.utilization
  } : null;
}
export { clearWorkerCache, destroyPool, extractToClassNames, extractToNative, getPoolStats, getPragmaOptions, loadTamagui, loadTamaguiBuildConfig, watchTamaguiConfig };
