// @tamagui/react-native-web-lite@2.0.0-rc.36/dist/cjs/vendor/react-native/VirtualizedList/VirtualizedListProps.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var VirtualizedListProps_exports = {};
__export(VirtualizedListProps_exports, {
  default: () => VirtualizedListProps_default,
  defaultVirtualizedListProps: () => defaultVirtualizedListProps
});
module.exports = __toCommonJS(VirtualizedListProps_exports);
const defaultVirtualizedListProps = {
  debug: !1,
  disableVirtualization: !1,
  extraData: null,
  getItemLayout: null,
  horizontal: !1,
  initialNumToRender: 10,
  initialScrollIndex: null,
  inverted: !1,
  keyExtractor: null,
  CellRendererComponent: null,
  ItemSeparatorComponent: null,
  ListItemComponent: null,
  ListEmptyComponent: null,
  ListFooterComponent: null,
  ListFooterComponentStyle: null,
  ListHeaderComponent: null,
  ListHeaderComponentStyle: null,
  maxToRenderPerBatch: 10,
  onEndReached: null,
  onEndReachedThreshold: 2,
  onRefresh: null,
  onScrollToIndexFailed: null,
  onStartReached: null,
  onStartReachedThreshold: 2,
  onViewableItemsChanged: null,
  persistentScrollbar: !1,
  progressViewOffset: 0,
  refreshControl: null,
  refreshing: !1,
  removeClippedSubviews: !1,
  renderScrollComponent: null,
  updateCellsBatchingPeriod: 50,
  viewabilityConfig: null,
  viewabilityConfigCallbackPairs: null,
  windowSize: 21
};
var VirtualizedListProps_default = defaultVirtualizedListProps;
