// @tamagui/react-native-web-lite@2.0.0-rc.36/dist/cjs/ScrollView/ScrollViewBase.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf,
  __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
    value: mod,
    enumerable: !0
  }) : target, mod)),
  __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
    value: !0
  }), mod);
var ScrollViewBase_exports = {};
__export(ScrollViewBase_exports, {
  ScrollViewBase: () => ScrollViewBase,
  default: () => ScrollViewBase_default
});
module.exports = __toCommonJS(ScrollViewBase_exports);
var import_react_native_web_internals = require("@tamagui/react-native-web-internals"),
  React = __toESM(require("react"), 1),
  import_View = require("../View/index.cjs"),
  import_jsx_runtime = require("react/jsx-runtime");
function normalizeScrollEvent(e) {
  return {
    nativeEvent: {
      contentOffset: {
        get x() {
          return e.target.scrollLeft;
        },
        get y() {
          return e.target.scrollTop;
        }
      },
      contentSize: {
        get height() {
          return e.target.scrollHeight;
        },
        get width() {
          return e.target.scrollWidth;
        }
      },
      layoutMeasurement: {
        get height() {
          return e.target.offsetHeight;
        },
        get width() {
          return e.target.offsetWidth;
        }
      }
    },
    timeStamp: Date.now()
  };
}
function shouldEmitScrollEvent(lastTick, eventThrottle) {
  const timeSinceLastTick = Date.now() - lastTick;
  return eventThrottle > 0 && timeSinceLastTick >= eventThrottle;
}
const ScrollViewBase = React.forwardRef((props, forwardedRef) => {
    const {
        onScroll,
        onTouchMove,
        onWheel,
        scrollEnabled = !0,
        scrollEventThrottle = 0,
        showsHorizontalScrollIndicator,
        showsVerticalScrollIndicator,
        style,
        // strip RN-only props that shouldn't reach the DOM
        onMomentumScrollBegin,
        onMomentumScrollEnd,
        onScrollBeginDrag,
        onScrollEndDrag,
        ...rest
      } = props,
      scrollState = React.useRef({
        isScrolling: !1,
        scrollLastTick: 0
      }),
      scrollTimeout = React.useRef(null),
      scrollRef = React.useRef(null);
    function createPreventableScrollHandler(handler) {
      return e => {
        scrollEnabled && handler && handler(e);
      };
    }
    function handleScroll(e) {
      e.stopPropagation(), e.target === scrollRef.current && (e.persist(), scrollTimeout.current != null && clearTimeout(scrollTimeout.current), scrollTimeout.current = setTimeout(() => {
        handleScrollEnd(e);
      }, 100), scrollState.current.isScrolling ? shouldEmitScrollEvent(scrollState.current.scrollLastTick, scrollEventThrottle) && handleScrollTick(e) : handleScrollStart(e));
    }
    function handleScrollStart(e) {
      scrollState.current.isScrolling = !0, handleScrollTick(e);
    }
    function handleScrollTick(e) {
      scrollState.current.scrollLastTick = Date.now(), onScroll && onScroll(normalizeScrollEvent(e));
    }
    function handleScrollEnd(e) {
      scrollState.current.isScrolling = !1, onScroll && onScroll(normalizeScrollEvent(e));
    }
    const hideHorizontalScrollbar = showsHorizontalScrollIndicator === !1,
      hideVerticalScrollbar = showsVerticalScrollIndicator === !1;
    return /* @__PURE__ */(0, import_jsx_runtime.jsx)(import_View.View, {
      ...rest,
      className: "_dsp_contents" + (hideHorizontalScrollbar ? " _hsb-x" : "") + (hideVerticalScrollbar ? " _hsb-y" : ""),
      onScroll: handleScroll,
      onTouchMove: createPreventableScrollHandler(onTouchMove),
      onWheel: createPreventableScrollHandler(onWheel),
      ref: (0, import_react_native_web_internals.useMergeRefs)(scrollRef, forwardedRef),
      style: [style,
      // @ts-ignore
      !scrollEnabled && styles.scrollDisabled]
    });
  }),
  styles = {
    scrollDisabled: {
      overflowX: "hidden",
      overflowY: "hidden",
      touchAction: "none"
    }
  };
var ScrollViewBase_default = ScrollViewBase;
