// @tamagui/vite-plugin@2.0.0-rc.36/dist/cjs/loadTamagui.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf,
  __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
    value: mod,
    enumerable: !0
  }) : target, mod)),
  __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
    value: !0
  }), mod);
var loadTamagui_exports = {};
__export(loadTamagui_exports, {
  cleanup: () => cleanup,
  ensureFullConfigLoaded: () => ensureFullConfigLoaded,
  getLoadPromise: () => getLoadPromise,
  getTamaguiOptions: () => getTamaguiOptions,
  loadTamaguiBuildConfig: () => loadTamaguiBuildConfig
});
module.exports = __toCommonJS(loadTamagui_exports);
var StaticWorker = __toESM(require("@tamagui/static-worker"), 1);
const LOAD_STATE_KEY = "__tamagui_load_state__";
function getLoadState() {
  return globalThis[LOAD_STATE_KEY] || (globalThis[LOAD_STATE_KEY] = {
    loadPromise: null,
    loadedOptions: null,
    fullConfigLoaded: !1,
    fullConfigLoadPromise: null
  }), globalThis[LOAD_STATE_KEY];
}
function getTamaguiOptions() {
  return getLoadState().loadedOptions;
}
function getLoadPromise() {
  return getLoadState().loadPromise;
}
async function loadTamaguiBuildConfig(optionsIn) {
  const state = getLoadState();
  return state.loadedOptions ? state.loadedOptions : (state.loadPromise || (state.loadPromise = (async () => {
    const options = await StaticWorker.loadTamaguiBuildConfig({
      ...optionsIn,
      platform: "web"
    });
    return state.loadedOptions = options, options;
  })()), state.loadPromise);
}
async function ensureFullConfigLoaded() {
  const state = getLoadState();
  if (!state.fullConfigLoaded) return state.fullConfigLoadPromise || (state.fullConfigLoadPromise = (async () => {
    const options = await loadTamaguiBuildConfig();
    !options.disableWatchTamaguiConfig && !options.disable && (await StaticWorker.loadTamagui({
      components: ["tamagui"],
      platform: "web",
      ...options
    })), state.fullConfigLoaded = !0;
  })()), state.fullConfigLoadPromise;
}
async function cleanup() {
  await StaticWorker.destroyPool();
  const state = getLoadState();
  state.loadPromise = null, state.loadedOptions = null, state.fullConfigLoaded = !1, state.fullConfigLoadPromise = null;
}
