// @tamagui/react-native-web-lite@2.0.0-rc.36/dist/esm/vendor/react-native/VirtualizedList/VirtualizedListProps.mjs
const defaultVirtualizedListProps = {
  debug: !1,
  disableVirtualization: !1,
  extraData: null,
  getItemLayout: null,
  horizontal: !1,
  initialNumToRender: 10,
  initialScrollIndex: null,
  inverted: !1,
  keyExtractor: null,
  CellRendererComponent: null,
  ItemSeparatorComponent: null,
  ListItemComponent: null,
  ListEmptyComponent: null,
  ListFooterComponent: null,
  ListFooterComponentStyle: null,
  ListHeaderComponent: null,
  ListHeaderComponentStyle: null,
  maxToRenderPerBatch: 10,
  onEndReached: null,
  onEndReachedThreshold: 2,
  onRefresh: null,
  onScrollToIndexFailed: null,
  onStartReached: null,
  onStartReachedThreshold: 2,
  onViewableItemsChanged: null,
  persistentScrollbar: !1,
  progressViewOffset: 0,
  refreshControl: null,
  refreshing: !1,
  removeClippedSubviews: !1,
  renderScrollComponent: null,
  updateCellsBatchingPeriod: 50,
  viewabilityConfig: null,
  viewabilityConfigCallbackPairs: null,
  windowSize: 21
};
var VirtualizedListProps_default = defaultVirtualizedListProps;
export { VirtualizedListProps_default as default, defaultVirtualizedListProps };
