// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/amberA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var amberA_exports = {};
__export(amberA_exports, {
  amberA: () => amberA
});
module.exports = __toCommonJS(amberA_exports);
const amberA = {
  amberA1: "#c0800004",
  amberA2: "#ffaa0012",
  amberA3: "#ffb3002a",
  amberA4: "#ffaa0043",
  amberA5: "#ffa6005d",
  amberA6: "#f49c0079",
  amberA7: "#e793009c",
  amberA8: "#e48b00d4",
  amberA9: "#ffb300dc",
  amberA10: "#f9a600e3",
  amberA11: "#ab5500fa",
  amberA12: "#341500dd"
};
