// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/blueA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var blueA_exports = {};
__export(blueA_exports, {
  blueA: () => blueA
});
module.exports = __toCommonJS(blueA_exports);
const blueA = {
  blueA1: "#0080ff04",
  blueA2: "#008cff0b",
  blueA3: "#008ff519",
  blueA4: "#009eff2a",
  blueA5: "#0093ff3d",
  blueA6: "#0088f653",
  blueA7: "#0083eb71",
  blueA8: "#0084e6a1",
  blueA9: "#0090ff",
  blueA10: "#0086f0fa",
  blueA11: "#006dcbf2",
  blueA12: "#002359ee"
};
