// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/brownA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var brownA_exports = {};
__export(brownA_exports, {
  brownA: () => brownA
});
module.exports = __toCommonJS(brownA_exports);
const brownA = {
  brownA1: "#aa550003",
  brownA2: "#aa550009",
  brownA3: "#a04b0015",
  brownA4: "#9b4a0022",
  brownA5: "#9f4a0033",
  brownA6: "#9a46004a",
  brownA7: "#99440069",
  brownA8: "#a0470090",
  brownA9: "#823300a7",
  brownA10: "#7424009e",
  brownA11: "#5d1c00a8",
  brownA12: "#140600d1"
};
