// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/crimsonA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var crimsonA_exports = {};
__export(crimsonA_exports, {
  crimsonA: () => crimsonA
});
module.exports = __toCommonJS(crimsonA_exports);
const crimsonA = {
  crimsonA1: "#ff005503",
  crimsonA2: "#e0004008",
  crimsonA3: "#ff005216",
  crimsonA4: "#f8005123",
  crimsonA5: "#e5004f31",
  crimsonA6: "#d0004b41",
  crimsonA7: "#bf004753",
  crimsonA8: "#b6004a6c",
  crimsonA9: "#e2005bc2",
  crimsonA10: "#d70056cb",
  crimsonA11: "#c4004fe2",
  crimsonA12: "#530026e9"
};
