// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/cyanA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var cyanA_exports = {};
__export(cyanA_exports, {
  cyanA: () => cyanA
});
module.exports = __toCommonJS(cyanA_exports);
const cyanA = {
  cyanA1: "#0099cc05",
  cyanA2: "#009db10d",
  cyanA3: "#00c2d121",
  cyanA4: "#00bcd435",
  cyanA5: "#01b4cc4a",
  cyanA6: "#00a7c162",
  cyanA7: "#009fbb82",
  cyanA8: "#00a3c0c2",
  cyanA9: "#00a2c7",
  cyanA10: "#0094b7f8",
  cyanA11: "#007491ef",
  cyanA12: "#00323ef2"
};
