// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/goldA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var goldA_exports = {};
__export(goldA_exports, {
  goldA: () => goldA
});
module.exports = __toCommonJS(goldA_exports);
const goldA = {
  goldA1: "#55550003",
  goldA2: "#a4840d0d",
  goldA3: "#91700016",
  goldA4: "#84620022",
  goldA5: "#7756002f",
  goldA6: "#6f4d0042",
  goldA7: "#6644005b",
  goldA8: "#63420e7d",
  goldA9: "#5320009a",
  goldA10: "#4b1d00a2",
  goldA11: "#3c1500af",
  goldA12: "#130c00d4"
};
