// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/grassA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var grassA_exports = {};
__export(grassA_exports, {
  grassA: () => grassA
});
module.exports = __toCommonJS(grassA_exports);
var grassA = {
  grassA1: "#00c00004",
  grassA2: "#00c0000c",
  grassA3: "#00b70014",
  grassA4: "#00a30020",
  grassA5: "#009a0031",
  grassA6: "#00900048",
  grassA7: "#008b0068",
  grassA8: "#00920e9a",
  grassA9: "#008619b9",
  grassA10: "#007b17c2",
  grassA11: "#006514d6",
  grassA12: "#002006e4"
};
