// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/greenA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var greenA_exports = {};
__export(greenA_exports, {
  greenA: () => greenA
});
module.exports = __toCommonJS(greenA_exports);
var greenA = {
  greenA1: "#00c04004",
  greenA2: "#00c03e0d",
  greenA3: "#00b83a16",
  greenA4: "#019c3522",
  greenA5: "#00963433",
  greenA6: "#008f304b",
  greenA7: "#00872e6d",
  greenA8: "#009437a4",
  greenA9: "#008f2fcf",
  greenA10: "#00832bd6",
  greenA11: "#006b23e6",
  greenA12: "#002616e6"
};
