// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/indigoA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var indigoA_exports = {};
__export(indigoA_exports, {
  indigoA: () => indigoA
});
module.exports = __toCommonJS(indigoA_exports);
const indigoA = {
  indigoA1: "#00008002",
  indigoA2: "#0040ff08",
  indigoA3: "#0047f112",
  indigoA4: "#0044ff1e",
  indigoA5: "#0044ff2d",
  indigoA6: "#003eff3e",
  indigoA7: "#0037ed54",
  indigoA8: "#0034dc72",
  indigoA9: "#0031d2c1",
  indigoA10: "#002ec9cc",
  indigoA11: "#002bb7c5",
  indigoA12: "#001046e0"
};
