// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/mauveA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var mauveA_exports = {};
__export(mauveA_exports, {
  mauveA: () => mauveA
});
module.exports = __toCommonJS(mauveA_exports);
const mauveA = {
  mauveA1: "#55005503",
  mauveA2: "#2b005506",
  mauveA3: "#30004010",
  mauveA4: "#20003618",
  mauveA5: "#20003820",
  mauveA6: "#14003527",
  mauveA7: "#10003332",
  mauveA8: "#08003145",
  mauveA9: "#05001d73",
  mauveA10: "#0500197d",
  mauveA11: "#0400119c",
  mauveA12: "#020008e0"
};
