// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/orangeA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var orangeA_exports = {};
__export(orangeA_exports, {
  orangeA: () => orangeA
});
module.exports = __toCommonJS(orangeA_exports);
const orangeA = {
  orangeA1: "#c0400004",
  orangeA2: "#ff8e000b",
  orangeA3: "#ff800018",
  orangeA4: "#ff7f0028",
  orangeA5: "#ff73003c",
  orangeA6: "#ff6a0058",
  orangeA7: "#ff62007e",
  orangeA8: "#ff5c00b2",
  orangeA9: "#ff5d00f7",
  orangeA10: "#fe5000f9",
  orangeA11: "#cf4100fa",
  orangeA12: "#431200e2"
};
