// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/purpleA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var purpleA_exports = {};
__export(purpleA_exports, {
  purpleA: () => purpleA
});
module.exports = __toCommonJS(purpleA_exports);
const purpleA = {
  purpleA1: "#aa00aa03",
  purpleA2: "#8000e008",
  purpleA3: "#8e00f112",
  purpleA4: "#8d00e51d",
  purpleA5: "#8000db2a",
  purpleA6: "#7a01d03b",
  purpleA7: "#6d00c350",
  purpleA8: "#6600c06c",
  purpleA9: "#5c00adb1",
  purpleA10: "#53009eb8",
  purpleA11: "#52009aba",
  purpleA12: "#250049df"
};
