// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/redA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var redA_exports = {};
__export(redA_exports, {
  redA: () => redA
});
module.exports = __toCommonJS(redA_exports);
const redA = {
  redA1: "#ff000003",
  redA2: "#ff000008",
  redA3: "#f3000d14",
  redA4: "#ff000824",
  redA5: "#ff000632",
  redA6: "#f8000442",
  redA7: "#df000356",
  redA8: "#d2000571",
  redA9: "#db0007b7",
  redA10: "#d10005c1",
  redA11: "#c40006d3",
  redA12: "#55000de8"
};
