// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/slateA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var slateA_exports = {};
__export(slateA_exports, {
  slateA: () => slateA
});
module.exports = __toCommonJS(slateA_exports);
const slateA = {
  slateA1: "#00005503",
  slateA2: "#00005506",
  slateA3: "#0000330f",
  slateA4: "#00002d17",
  slateA5: "#0009321f",
  slateA6: "#00002f26",
  slateA7: "#00062e32",
  slateA8: "#00083046",
  slateA9: "#00051d74",
  slateA10: "#00071b7f",
  slateA11: "#0007149f",
  slateA12: "#000509e3"
};
