// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/violetA.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var violetA_exports = {};
__export(violetA_exports, {
  violetA: () => violetA
});
module.exports = __toCommonJS(violetA_exports);
const violetA = {
  violetA1: "#5500aa03",
  violetA2: "#4900ff07",
  violetA3: "#4400ee0f",
  violetA4: "#4300ff1b",
  violetA5: "#3600ff26",
  violetA6: "#3100fb35",
  violetA7: "#2d01dd4a",
  violetA8: "#2b00d066",
  violetA9: "#2400b7a9",
  violetA10: "#2300abb2",
  violetA11: "#1f0099af",
  violetA12: "#0b0043d9"
};
