// @tamagui/colors@2.0.0-rc.36/dist/cjs/light/yellowA.native.js
"use strict";

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
    for (var name in all) __defProp(target, name, {
      get: all[name],
      enumerable: !0
    });
  },
  __copyProps = (to, from, except, desc) => {
    if (from && typeof from == "object" || typeof from == "function") for (let key of __getOwnPropNames(from)) !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    });
    return to;
  };
var __toCommonJS = mod => __copyProps(__defProp({}, "__esModule", {
  value: !0
}), mod);
var yellowA_exports = {};
__export(yellowA_exports, {
  yellowA: () => yellowA
});
module.exports = __toCommonJS(yellowA_exports);
var yellowA = {
  yellowA1: "#aaaa0006",
  yellowA2: "#c4a80017",
  yellowA3: "#c2a10a2e",
  yellowA4: "#c09b0044",
  yellowA5: "#b896005b",
  yellowA6: "#ae8d0373",
  yellowA7: "#a38100a0",
  yellowA8: "#997700f6",
  yellowA9: "#a89200f5",
  yellowA10: "#a48d00f7",
  yellowA11: "#724f00fb",
  yellowA12: "#2e2000e0"
};
